Add CentOS 8 image dockerfile

avtikhon opened the pull request:
Added the new docker file for CentOS 8, checked with Packpack and Tarantool build.

avtikhon committed:
Add CentOS 8 image dockerfile

Added CentOS 8 image named as el-8 with the following differences from
CentOS 7:
(check https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/considerations_in_adopting_rhel_8/index )
 - added PowerTools repository as suggested at
   https://fedoraproject.org/wiki/EPEL
 - removed curl which is allready installed at the parent image;
 - at CentOS 8 the localization environment changed to "C";
   (needed by backport repository);
 - removed pygpgme which was removed at CentOS 8 (check A.3.
   "Removed packages part");
 - saved backport repository due to python2 still in use while CentOS 8
   changed its packages to python3 and we need to backport python-gevent
   package;
 - cmake28 and cmake3 removed due to CentOS 8 installs 3.11 version;
 - removed devtoolset and scl* repositories due to devtoolset-8 is
   already in the standard repositories and scl* repositories still
   not available.

akopytov commented:
Dockerfile being added in this PR doesn't work for me due to gevent (which doesn't seem to be present in the images for CentOS 6 and 7?). This is how it fails for me when I run `make centos-8`:

```console
WARNING: Running pip install with root privileges is generally not a good idea. Try `pip2 install --user` instead.
Collecting gevent
  Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ff25ac91d50>: Failed to establish a new connection: [Errno 111] Connection refused',)': /packages/e6/e9/3a693414f706e12abe60554cd73c5ae8f848b182ae58018f93d86c9eb418/gevent-1.4.0-cp27-cp27mu-manylinux1_x86_64.whl
  Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ff25ac87990>: Failed to establish a new connection: [Errno 111] Connection refused',)': /packages/e6/e9/3a693414f706e12abe60554cd73c5ae8f848b182ae58018f93d86c9eb418/gevent-1.4.0-cp27-cp27mu-manylinux1_x86_64.whl
  Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ff25ac9b950>: Failed to establish a new connection: [Errno 111] Connection refused',)': /packages/e6/e9/3a693414f706e12abe60554cd73c5ae8f848b182ae58018f93d86c9eb418/gevent-1.4.0-cp27-cp27mu-manylinux1_x86_64.whl
  Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ff25aca5550>: Failed to establish a new connection: [Errno 111] Connection refused',)': /packages/e6/e9/3a693414f706e12abe60554cd73c5ae8f848b182ae58018f93d86c9eb418/gevent-1.4.0-cp27-cp27mu-manylinux1_x86_64.whl
  Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ff25aca5690>: Failed to establish a new connection: [Errno 111] Connection refused',)': /packages/e6/e9/3a693414f706e12abe60554cd73c5ae8f848b182ae58018f93d86c9eb418/gevent-1.4.0-cp27-cp27mu-manylinux1_x86_64.whl
Exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python2.7/site-packages/pip/commands/install.py", line 346, in run
    requirement_set.prepare_files(finder)
  File "/usr/lib/python2.7/site-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/usr/lib/python2.7/site-packages/pip/req/req_set.py", line 620, in _prepare_file
    session=self.session, hashes=hashes)
  File "/usr/lib/python2.7/site-packages/pip/download.py", line 821, in unpack_url
    hashes=hashes
  File "/usr/lib/python2.7/site-packages/pip/download.py", line 659, in unpack_http_url
    hashes)
  File "/usr/lib/python2.7/site-packages/pip/download.py", line 853, in _download_http_url
    stream=True,
  File "/usr/lib/python2.7/site-packages/pip/_vendor/requests/sessions.py", line 521, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python2.7/site-packages/pip/download.py", line 386, in request
    return super(PipSession, self).request(method, url, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pip/_vendor/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/pip/_vendor/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/pip/_vendor/cachecontrol/adapter.py", line 47, in send
    resp = super(CacheControlAdapter, self).send(request, **kw)
  File "/usr/lib/python2.7/site-packages/pip/_vendor/requests/adapters.py", line 508, in send
    raise ConnectionError(e, request=request)
ConnectionError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Max retries exceeded with url: /packages/e6/e9/3a693414f706e12abe60554cd73c5ae8f848b182ae58018f93d86c9eb418/gevent-1.4.0-cp27-cp27mu-manylinux1_x86_64.whl (Caused by NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ff25aca5790>: Failed to establish a new connection: [Errno 111] Connection refused',))
The command '/bin/sh -c pip2 install gevent' returned a non-zero code: 2
```

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+FROM centos:8
+MAINTAINER Roman Tsisyk <roman@tarantool.org>
```
Please, set youself as the maintainer for newly created docker files.

avtikhon replied:
Updated.

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+MAINTAINER Roman Tsisyk <roman@tarantool.org>
+
+# Fix missing locales
+ENV LANGUAGE="en_US.UTF-8"
```
We set LC_ALL and LANG here for CentOS 7. Can you clarify this change in the commit message (short description + proper link is okay)?

avtikhon replied:
Added comment.

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+    wget \
+    yum-utils \
+    epel-release
+RUN curl -s https://packagecloud.io/install/repositories/packpack/backports/script.rpm.sh | bash
```
We have no backports even for CentOS 7. Maybe this is good time to remove using of this repository.

avtikhon replied:
Found that all packages at CentOS 8 moved from python2 to python3 and it is needed to store the backported packages with python2 for testing.

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+
+# Enable extra repositories
+RUN yum -y install \
+    wget \
```
We install curl and pygpgme on CentOS 7. As far as I see curl is installed in a parent image, so we can skip it (but it worth to leave a word about it in the commit message). The same for pygpgme: I guess it was a dependency for some of our scripts and is not needed anymore. It does not exists on CentOS 8 and seems to be deprecated in favor of python2-gpg (I [found](https://bugs.launchpad.net/pygpgme/+bug/1800660) the related discussion in pygpgme issue tracker). The same: it worth to mention this is the commit message.

avtikhon replied:
Added comment:
- removed curl which is allready installed at the parent image
- removed pygpgme which is also removed at CentOS 8 (check A.3.  "Removed packages part");
 https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/considerations_in_adopting_rhel_8/index

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+# Install base toolset
+RUN yum -y groupinstall 'Development Tools'
+RUN yum -y install \
+    cmake \
```
The new CentOS does not provide cmake28 and cmake3 anymore? Okay, but mention it in the commit message.

avtikhon replied:
Added comment: cmake28 and cmake3 removed due to CentOS 8 installs 3.11 version.

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+# Install base toolset
+RUN yum -y groupinstall 'Development Tools'
+RUN yum -y install \
+    cmake \
```
You removed devtoolset (as well as centos-release-scl centos-release-scl-rh repos). What is a reason? Please, clarify in the commit message too.

avtikhon replied:
Added comment: CentOS 8 doesn't have it at all and devtoolset-8 is already in the standard repositories and scl* repositories still not available.

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+
+# Install python2 for testing
+RUN dnf -y install python2
+RUN pip2 install gevent
```
We don't install neither python2, nor gevent in packpack/packpack:centos-7 explicitly. Why we doing it here? Please, re-check whether it is actually necessary. I think that we should keep in the mind that CentOS 8 announcements declare that neither python2, nor python3 is installed by default: we should follow this rule for such general-purpose image.

avtikhon replied:
Removed.

Totktonada reviewed (commented):
@avtikhon Thanks! Please, consider comments below.

@akopytov Cannot reproduce your problem. Maybe it was a temporary infrastructure problem?

akopytov commented:
> @akopytov Cannot reproduce your problem. Maybe it was a temporary infrastructure problem?

@Totktonada confirmed, it was a temporary problem. It works fine now, but I also don't understand the purpose of those extra packages, so I removed those lines in my branch.

Totktonada commented:
@akopytov Which lines you are saying about?

avtikhon commented:
>Вторник, 22 октября 2019, 19:03 +03:00 от Alexander Turenko <notifications@github.com>:
>
>@Totktonada commented on this pull request.
>@avtikhon Thanks! Please, consider comments below.
>@akopytov Cannot reproduce your problem. Maybe it was a temporary infrastructure problem?
>----------------------------------------------------------------------
>In  centos/8/Dockerfile :
>> @@ -0,0 +1,25 @@
>+FROM centos:8
>+MAINTAINER Roman Tsisyk <roman@tarantool.org>
>
>Please, set youself as the maintainer for newly created docker files. 
Updated.
>----------------------------------------------------------------------
>In  centos/8/Dockerfile :
>> @@ -0,0 +1,25 @@
>+FROM centos:8
>+MAINTAINER Roman Tsisyk <roman@tarantool.org>
>+
>+# Fix missing locales
>+ENV LANGUAGE="en_US.UTF-8"
>
>We set LC_ALL and LANG here for CentOS 7. Can you clarify this change in the commit message (short description + proper link is okay)? 
Added comment.
>----------------------------------------------------------------------
>In  centos/8/Dockerfile :
>> @@ -0,0 +1,25 @@
>+FROM centos:8
>+MAINTAINER Roman Tsisyk <roman@tarantool.org>
>+
>+# Fix missing locales
>+ENV LANGUAGE="en_US.UTF-8"
>+
>+# Enable extra repositories
>+RUN yum -y install \
>+    wget \
>+    yum-utils \
>+    epel-release
>+RUN curl -s https://packagecloud.io/install/repositories/packpack/backports/script.rpm.sh | bash
>
>We have no backports even for CentOS 7. Maybe this is good time to remove using of this repository. 
Found that all packages at CentOS 8 moved from python2 to python3 and it is needed to store the backported packages with python2 for testing.
>----------------------------------------------------------------------
>In  centos/8/Dockerfile :
>> @@ -0,0 +1,25 @@
>+FROM centos:8
>+MAINTAINER Roman Tsisyk <roman@tarantool.org>
>+
>+# Fix missing locales
>+ENV LANGUAGE="en_US.UTF-8"
>+
>+# Enable extra repositories
>+RUN yum -y install \
>+    wget \
>
>We install curl and pygpgme on CentOS 7. As far as I see curl is installed in a parent image, so we can skip it (but it worth to leave a word about it in the commit message). The same for pygpgme: I guess it was a dependency for some of our scripts and is not needed anymore. It does not exists on CentOS 8 and seems to be deprecated in favor of python2-gpg (I  found the related discussion in pygpgme issue tracker). The same: it worth to mention this is the commit message. 
Added comment.
>----------------------------------------------------------------------
>In  centos/8/Dockerfile :
>> +MAINTAINER Roman Tsisyk <roman@tarantool.org>
>+
>+# Fix missing locales
>+ENV LANGUAGE="en_US.UTF-8"
>+
>+# Enable extra repositories
>+RUN yum -y install \
>+    wget \
>+    yum-utils \
>+    epel-release
>+RUN curl -s https://packagecloud.io/install/repositories/packpack/backports/script.rpm.sh | bash
>+
>+# Install base toolset
>+RUN yum -y groupinstall 'Development Tools'
>+RUN yum -y install \
>+    cmake \
>
>The new CentOS does not provide cmake28 and cmake3 anymore? Okay, but mention it in the commit message. 
Added comment.
>----------------------------------------------------------------------
>In  centos/8/Dockerfile :
>> +MAINTAINER Roman Tsisyk <roman@tarantool.org>
>+
>+# Fix missing locales
>+ENV LANGUAGE="en_US.UTF-8"
>+
>+# Enable extra repositories
>+RUN yum -y install \
>+    wget \
>+    yum-utils \
>+    epel-release
>+RUN curl -s https://packagecloud.io/install/repositories/packpack/backports/script.rpm.sh | bash
>+
>+# Install base toolset
>+RUN yum -y groupinstall 'Development Tools'
>+RUN yum -y install \
>+    cmake \
>
>You removed devtoolset (as well as centos-release-scl centos-release-scl-rh repos). What is a reason? Please, clarify in the commit message too. 
CentOS 8 doesn't have it at all and devtoolset-8 is already in the standard repositories and scl* repositories still not available.

>----------------------------------------------------------------------
>In  centos/8/Dockerfile :
>> +# Enable extra repositories
>+RUN yum -y install \
>+    wget \
>+    yum-utils \
>+    epel-release
>+RUN curl -s https://packagecloud.io/install/repositories/packpack/backports/script.rpm.sh | bash
>+
>+# Install base toolset
>+RUN yum -y groupinstall 'Development Tools'
>+RUN yum -y install \
>+    cmake \
>+    sudo
>+
>+# Install python2 for testing
>+RUN dnf -y install python2
>+RUN pip2 install gevent
>
>We don't install neither python2, nor gevent in packpack/packpack:centos-7 explicitly. Why we doing it here? Please, re-check whether it is actually necessary. I think that we should keep in the mind that CentOS 8 announcements declare that neither python2, nor python3 is installed by default: we should follow this rule for such general-purpose image. 
Removed.
>—
>You are receiving this because you were mentioned.
>Reply to this email directly,  view it on GitHub , or  unsubscribe .

-- 
Alexander Tikhonov

akopytov commented:
> @akopytov Which lines you are saying about?

@Totktonada the ones installing `python2` and `gevent`. Those lines have been removed by now, so it's all fine.

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+RUN yum -y install wget yum-utils
+
+# Enable extra repositories
+RUN yum -y install epel-release
```
It is [recommended](https://fedoraproject.org/wiki/EPEL) to enable PowerTools repository when enabling EPEL. AFAIU it is a collection of developer oriented packages. I also [found](https://centos.pkgs.org/8/centos-powertools-x86_64/libyaml-devel-0.1.7-5.el8.x86_64.rpm.html) that it contains libyaml-devel package that we'll need for building of tarantool (at least for 1.10 and 2.1, see [here](https://github.com/tarantool/doc/issues/960)).

I made a test: excluded libyaml-devel from build requirements in the tarantool RPM spec and enabled -DENABLE_BUNDLED_LIBYAML:BOOL=OFF back in the spec on tarantool master (to imitate 1.10/2.1 build) and got an error during cmake stage w/o libyaml-devel package.

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+# Enable extra repositories
+RUN yum -y install epel-release
+
+# Backport repository for packages like python2 tools which moved to python3
```
Proposed wording:

> Repository for building/testing dependencies that are not present in vanilla CentOS and PowerTools / EPEL repositories, e.g. some Python 2 packages.

avtikhon replied:
Changed:
     - at CentOS 8 the localization environment changed to "C";
       (needed by backport repository);
     - changed pygpgme to python3-gpg which was removed at CentOS 8 (check A.3.
       "Removed packages part") (needed by backport repository);
     - saved backport repository due to python2 still in use while CentOS 8
       changed its packages to python3 and we need to backport python-gevent
       package;

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# - fix missing locales
+ENV LC_ALL="C" LANG="en_US.UTF-8"
+# - add python3-gpg instead of pygpgme
+RUN yum -y install python3-gpg
```
I would not install it, because the packagecloud script does not use the package at the time (it tries to install pygpgme and give a warning). Even when it will do, it seems the script will install python3-gpg itself.

avtikhon replied:
Ok, removed

Totktonada commented on centos/8/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+RUN yum -y install epel-release
+# added PowerTools as suggested at:
+#   https://fedoraproject.org/wiki/EPEL
+RUN dnf config-manager --set-enabled PowerTools
```
I would use `yum` everywhere for consistency.

avtikhon replied:
Changed to yum.

Totktonada reviewed (commented):
Other then the comments below everything looks okay. I'll wait for uploading python2-gevent RPM for CentOS 8 to backports repo and then will verify everything again on our CentOS 8 packpack build.

Totktonada reviewed (approved):
LGTM.
